src/Reflection/Annotated.agda

```diff
@@ -0,0 +1,335 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Annotated reflected syntax.
+--
+-- NOTE: This file does not check under --without-K due to restrictions
+--       in the termination checker. In particular recursive functions
+--       over a universe of types is not supported by --without-K.
+------------------------------------------------------------------------
+
+{-# OPTIONS --safe --with-K #-}
+
+module Reflection.Annotated where
+
+open import Level                               using (Level; 0ℓ; suc; _⊔_)
+open import Category.Applicative                using (RawApplicative)
+open import Data.Bool                           using (Bool; false; true; if_then_else_)
+open import Data.List                           using (List; []; _∷_)
+open import Data.List.Relation.Unary.All as All using (All; _∷_; [])
+open import Data.Product                        using (_×_; _,_; proj₁; proj₂)
+open import Data.String                         using (String)
+
+open import Reflection
+open import Reflection.Universe
+
+open Clause
+open Pattern
+open Sort
+
+------------------------------------------------------------------------
+-- Annotations and annotated types
+
+-- An Annotation is a type indexed by a reflected term.
+Annotation : ∀ ℓ → Set (suc ℓ)
+Annotation ℓ = ∀ {u} → ⟦ u ⟧ → Set ℓ
+
+-- An annotated type is a family over an Annotation and a reflected term.
+Typeₐ : ∀ ℓ → Univ → Set (suc ℓ)
+Typeₐ ℓ u = Annotation ℓ → ⟦ u ⟧ → Set ℓ
+
+private
+  variable
+    ℓ             : Level
+    u             : Univ
+    t             : ⟦ u ⟧
```

Comment:
At the moment the style guide for the library says that only levels and sets should be used as variables.

Reply:
I would argue for updating the style guide. Is there a reason for limiting it to levels and sets? In this case it will make the data type definitions a lot more verbose. Compare

```agda
-- with variables
data Argₐ′ (Tyₐ : Typeₐ ℓ u) : Typeₐ ℓ (⟨arg⟩ u) where
  arg : ∀ i → Tyₐ Ann t → Argₐ′ Tyₐ Ann (arg i t)

-- without variables
data Argₐ′ {u} (Tyₐ : Typeₐ ℓ u) : Typeₐ ℓ (⟨arg⟩ u) where
  arg : ∀ {Ann : Annotation ℓ} {t : ⟦ u ⟧} i → Tyₐ Ann t → Argₐ′ Tyₐ Ann (arg i t)
```

Reply:
Hmm looking back at the discussion in #697, we didn't want people going overboard with variables at the expense of readability. In the back of my mind at the time was also the concern that variables were very new, and therefore _may_ have had bugs in so perhaps we didn't want to jump in to using them too deeply.

The latter concern has obviously been addressed by time, but the former is still valid. The problem is that to the author of a piece of code, it's blindingly obvious what the types of variables are, but to readers it may be much more difficult to suss out. I guess it's also partly me being lazy and not wanting to have to get into an argument about readability in every PR that contains an overzealous use of variables.

What are other people's thoughts? @JacquesCarette? @gallais?

Reply:
If you think it improves readability I think it's fair to drop the `t` variable. I don't see any readability argument for `u` since without the variable there would just be a bunch of `∀ {u}` everywhere instead (and it plays the role of a Set here, so could be covered by the "only sets and levels" restriction). The `Ann` variable removes a lot of noise from the primed datatypes, and since it's a central concept of the module I would argue that it should not lead to any confusion to make it a variable.

Reply:
I think we could extend the criterion to "values in the module's universe of discourse".
So `Nat` variables in `Data.Nat.Properties`, `List` ones in `Data.List.Relation.Unary.Any`
and Reflection's `Universe` ones in `Reflection`.

This way we still have a clear criterion to avoid infinite debates but are a bit more
liberal when the context ought to make what these are obvious.

Reply:
> If you think it improves readability I think it's fair to drop the t variable. I don't see any readability argument for u since without the variable there would just be a bunch of ∀ {u} everywhere instead (and it plays the role of a Set here, so could be covered by the "only sets and levels" restriction). The Ann variable removes a lot of noise from the primed datatypes, and since it's a central concept of the module I would argue that it should not lead to any confusion to make it a variable.

> I think we could extend the criterion to "values in the module's universe of discourse".
So Nat variables in Data.Nat.Properties, List ones in Data.List.Relation.Unary.Any
and Reflection's Universe ones in Reflection.

Both of these seem reasonable to me :+1:

Reply:
Sorry for the slowness in answer. I agree with @gallais 's suggestion on "values in the module's universe of discourse".

However, I am still a tiny bit wary: in agda-categories, we had modules that quantified over arrows that broke because different versions of Agda inserted the parameters in different orders, and that led to problems: you can't reliably specify those, even by name, at call sites!

So the rule perhaps ought to be "you can use a variable to abstract over a parameter that is in the domain of discourse of the module AND you never expect that variable to be hand-instantiated in user code".  At least, until Agda treats its named telescopes more DAG-like than as strict lists.